crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -471,25 +503,35 @@ module sui_system::validator_set {
 
     fun update_and_process_low_stake_departures(
         self: &mut ValidatorSet,
-        low_stake_threshold: u64,
-        very_low_stake_threshold: u64,
         low_stake_grace_period: u64,
         validator_report_records: &mut VecMap<address, VecSet<address>>,
         ctx: &mut TxContext
     ) {
+        let initial_total_stake = calculate_total_stakes(&self.active_validators);
+        // prior to this line, `self.total_stake` is the previous epoch's total stake. we want to update it to the current total stake
+        // before processing low stake departures, otherwise the `total_stake` denominator in our voting power calculations will be incorrect.
+        // note: if a validator is kicked out in the loop below, `total_stake` will be decreased by `process_validator_departure`. otherwise,
+        // `total_stake` will be unchanged by the end of the function
```

Comment:
I noticed that in the previous implementation, we first triggered the processing of departures, then set the `total_stake` based on the "remaining" active set of validators (from scratch).

Here, we calculate it and set it before we process the departing validators, then allow the departure to deduct from it.

Are we sure this cannot cause any side-effect issues in any way? I do see that on the codepath, we do deduct the total_stake when we actually process a validator departure, however I thought I'd triple-check!

Reply:
I read through this and convinced myself that this way is indeed correct. But this made me realize that the statefulness of `total_stake` makes this code very hard to reason about indeed... I did some light refactoring to only write to `total_stake` in one location, and make sure that all logic in `validator_set` and `voting_power` is using the same value (e.g., there were separate functions in each module for computing total stake before)

Reply:
Thanks Sam. I think the change indeed makes it easier to follow! The consistency we get on `voting_power` is a great deal.